### PR TITLE
fix #664 possible race in ThreadId.setError method

### LINKED ISSUE #664: testInstallLargeSnapshotWithThrottle is flaky
### Describe the bug
Test `com.alipay.sofa.jraft.core.NodeTest#testInstallLargeSnapshotWithThrottle` is flaky. It is needed to run this test approxiamtely 20 times to get this error 

```
2021-08-11 19:16:57 [JRaft-Closure-Executor-7] ERROR Utils:180 - Fail to run done closure
java.lang.AssertionError: Status[ECATCHUP<10003>: Peer 172.25.4.112:5006 failed to catch up.]
	at org.junit.Assert.fail(Assert.java:89) ~[junit-4.13.1.jar:4.13.1]
	at org.junit.Assert.assertTrue(Assert.java:42) ~[junit-4.13.1.jar:4.13.1]
	at com.alipay.sofa.jraft.core.NodeTest.lambda$testInstallLargeSnapshotWithThrottle$1(NodeTest.java:2194) ~[test-classes/:?]
	at com.alipay.sofa.jraft.util.Utils$1.run(Utils.java:178) [classes/:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [?:1.8.0_231]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:1.8.0_231]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_231]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_231]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_231]
```

The first time we encountered this failed test in our fork, we changed RPC layer in our repo, so as far as we reproduced this also in the original JRaft repo, I would say that this bug is not related to the network layer. I've also made some investigations on our side, I've recorded messages on the leader and we can see that right before the fail leader for some reason sent last two `AppendEntriesRequest` with an interval that is more than election timeout. As far as I can understand, that affects `lastRpcSendTimestamp` and further check for a catch-up does not pass and ends with an error 
```
Utils.monotonicMs() - this.replicatorGroup.getLastRpcSendTimestamp(peer) <= this.options
                    .getElectionTimeoutMs()
```

I can provide full debug logs, If it is needed

### Expected behavior
Node is successfully joined to the cluster with throttled snapshot and the test is not flaky

### Actual behavior
Test `testInstallLargeSnapshotWithThrottle` is flaky, and there is a chance that node cannot join the topology successfully 

### Steps to reproduce
Just run `testInstallLargeSnapshotWithThrottle` several time (~20) on the last matser revision (34af4e21cf976f7d5499634bf74f5686c5a70677).

### Environment

- SOFAJRaft version: 1.3.8
- JVM version (e.g. `java -version`):
java version "1.8.0_231"
Java(TM) SE Runtime Environment (build 1.8.0_231-b11)
Java HotSpot(TM) 64-Bit Server VM (build 25.231-b11, mixed mode)
- OS version (e.g. `uname -a`):
macOS Catalina 10.15.7
- Maven version:
Apache Maven 3.6.3
- IDE version:
Intellij IDEA 2019.3.5

### PR DESCRIPTION
**Pull request by alievmirza**
Fix for https://github.com/sofastack/sofa-jraft/issues/664 
See details in the provided link

### CODE REVIEW

**Comment by killme2008** [comment]
@alievmirza  Please sign the CLA.

**Comment by alievmirza** [comment]
@killme2008 I checked sofastack-bot that is needed for CLA, and it seems to me that this bot asks me to grant access to some private organizations from my github account, is it required if I want to contribute to the project?

**Comment by seeflood** [comment]
> @killme2008 I checked sofastack-bot that is needed for CLA, and it seems to me that this bot asks me to grant access to some private organizations from my github account, is it required if I want to contribute to the project?

You are right,there is no need to grant so much privilege and I will fix the robot.
Now you can sign cla with smaller scope by changing the scope field in the url,check this   https://github.com/login/oauth/authorize?client_id=1bd0989ae4b5298d11b5&scope=user

**Comment by alievmirza** [comment]
@seeflood Thank you for the help! I've changed the scope and granted access only to my public info.

**Comment by seeflood** [comment]
@alievmirza 

Sorry for the inconvenience,I'm afraid that the cla was not signed. I just checked the cla database and found that there is no signature data for your cla.
Could you please re-sign cla?
Open http://cla.sofastack.tech/ and click "Sign Individual CLA".If it succeeds, a prompt will appear:




Maybe you just authorized to access the cla website through github oauth, but you did not click the button to sign cla after you authorized oauth?
Sorry again for the inconvenience.

**Comment by alievmirza** [comment]
@seeflood Thank you for the help! 

I used this link https://github.com/login/oauth/authorize?client_id=1bd0989ae4b5298d11b5 (without user scope, because it still asks access to my private organization). Now I can see my github logo on the top right of the site.
image
When I push Sign Individual CLA, it again redirects to bot authorization and asks for full access to my github account.

**Comment by seeflood** [comment]
@alievmirza 
I tried to use another account to sign the cla but can't reproduce your problem 😢 
My steps:
1. Open your link https://github.com/login/oauth/authorize?client_id=1bd0989ae4b5298d11b5
2. Grant access
3. Now it redirect to http://cla.sofastack.tech/
4. I clicked the "Sign Individual CLA" button
5. It redirect to a form page
6. I wrote all the information and submitted the form

I guess it's the unstable network between our cla server and github server that leads to your problem. 
Could you please try again?

**Comment by alievmirza** [comment]
@seeflood Revoking sofastack-bot app from my github account helped me and seems CLA is signed now. Thanks! 
@killme2008 Could you please review the changes?

File: jraft-core/src/main/java/com/alipay/sofa/jraft/util/ThreadId.java [hunk 1]
```diff
@@ -157,17 +157,19 @@ public void setError(final int errorCode) {
...
+                if (this.destroyed) {
+                    this.lock.unlock();
+                    return;
+                }
+                if (this.onError != null) {
+                    // The lock will be unlocked in onError.
+                    this.onError.onError(this, this.data, errorCode);
+                }
+            } else {
+                this.pendingErrors.add(errorCode);
```
Comment (fengjiachun) [inline jraft-core/src/main/java/com/alipay/sofa/jraft/util/ThreadId.java:171]:
```
private final List<Integer>    pendingErrors       = Collections.synchronizedList(new ArrayList<>());
```
Should we redeclare `pendingErrors` as An ArrayList?

Comment (alievmirza) [reply jraft-core/src/main/java/com/alipay/sofa/jraft/util/ThreadId.java:171]: done

**Review by killme2008 — APPROVED** [review]

**Comment by killme2008** [comment]
LGTM

**Review by fengjiachun — APPROVED** [review]
LGTM

**Comment by killme2008** [comment]
@alievmirza  thank u.

**Pull request merged**